src/lib/storage/dictionary_segment.hpp

```diff
@@ -25,51 +28,104 @@ class DictionarySegment : public BaseSegment {
   /**
    * Creates a Dictionary segment from a given value segment.
    */
-  explicit DictionarySegment(const std::shared_ptr<BaseSegment>& base_segment);
-
-  // SEMINAR INFORMATION: Since most of these methods depend on the template parameter, you will have to implement
-  // the DictionarySegment in this file. Replace the method signatures with actual implementations.
-
-  // return the value at a certain position. If you want to write efficient operators, back off!
-  AllTypeVariant operator[](const ChunkOffset chunk_offset) const override;
-
-  // return the value at a certain position.
-  T get(const size_t chunk_offset) const;
+  explicit DictionarySegment(const std::shared_ptr<BaseSegment>& base_segment) {
+    const auto segment_size = base_segment->size();
+    _dictionary = std::make_shared<std::vector<T>>(segment_size);
+
+    for (auto cell_id = 0u; cell_id < segment_size; ++cell_id) {
+      (*_dictionary)[cell_id] = type_cast<T>((*base_segment)[cell_id]);
+    }
+
+    std::sort(_dictionary->begin(), _dictionary->end());
+    const auto first_duplicate_element = std::unique(_dictionary->begin(), _dictionary->end());
+    _dictionary->erase(first_duplicate_element, _dictionary->end());
+    _dictionary->shrink_to_fit();
+
+    const auto dictionary_size = _dictionary->size();
+
+    // Calculate appropriate width for attribute vector
+    // by finding the number of bits needed to store the maximum valueID
+    const auto required_bit_size = std::ceil(std::log(dictionary_size) / std::log(2));
+
+    if (required_bit_size <= 8) {
+      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint8_t>>(segment_size);
+    } else if (required_bit_size <= 16) {
+      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint16_t>>(segment_size);
+    } else {
+      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint32_t>>(segment_size);
+    }
+
+    for (auto cell_id = 0u; cell_id < segment_size; ++cell_id) {
+      const auto value = type_cast<T>((*base_segment)[cell_id]);
+      const auto dictionary_offset =
+          std::distance(_dictionary->cbegin(), std::lower_bound(_dictionary->cbegin(), _dictionary->cend(), value));
+      (*_attribute_vector).set(cell_id, static_cast<ValueID>(dictionary_offset));
+    }
+  }
+
+  // returns the value at a certain position. If you want to write efficient operators, back off!
+  AllTypeVariant operator[](const ChunkOffset chunk_offset) const override {
+    const auto dictionary_value = _dictionary->at(_attribute_vector->get(chunk_offset));
+    return static_cast<AllTypeVariant>(dictionary_value);
```

Comment:
Consider using the function get() to avoid code duplication.
```suggestion
    return static_cast<AllTypeVariant>(get(chunk_offset));
```